src/test/org/codehaus/groovy/runtime/DefaultGroovyMethodsTest.groovy

```diff
@@ -236,7 +236,28 @@ public class DefaultGroovyMethodsTest extends GroovyTestCase {
         assertEquals(3, list.get(2));
     }
 
+    // GROOVY-7654
+    public void testIterableAsList() {
+        def list = [1, 2, 3]
+        def iterable = new IterableWrapper(delegate: list)
+
+        def iterableAsList = iterable.asList()
+        def iterableAsType = iterable as List
+        
+        assertEquals(iterableAsList, iterableAsType)
+        assertEquals(1, iterableAsList[0])
+        assertEquals(1, iterableAsType[0])
+    }
+
     private static class MyList extends ArrayList {
         public MyList() {}
     }
+
+    private static class IterableWrapper implements Iterable {
+        Iterable delegate
+
+        Iterator iterator() {
+            delegate.iterator()
+        }
+    }
 }
```

Comment:
Actually, test I got from jira issue and meaning I think is to create concrete implementation of Iterable which is not extending from collection. And for concrete implementation we need return some iterator.

Also I thinking to change code like:

```
public static <T> T asType(Iterable iterable, Class<T> clazz) {
        if (Collections.class.isAssignableFrom(clazz)) {
            return asType((Collection) toList(iterable), clazz);
        }

        return asType((Object) toList(iterable), clazz);
    }
```